File: cpp-solutions/33.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l=0, r=nums.size()-1;
        if(target<nums[l] && target>nums[r]) return -1;
        while(l<r){
            int m = (l+r)/2;
            if(nums[m]<=nums[r]){
                if(nums[m]<target && target<=nums[r]){
                    l = m+1;
                }else{
                    r = m;
                }
            }else{
                if(nums[l]<=target && target<=nums[m]){
                    r = m;
                }else{
                    l = m+1;
                }
            }
        }
        if(nums[l]==target) return l;
        return -1;
    }
};

File: go-solutions/33.go
package main

func search(nums []int, target int) int {
	if len(nums) == 0 {
		return -1
	}
	splitPos := 0
	for i := 1; i < len(nums); i++ {
		if nums[i] < nums[i-1] {
			splitPos = i - 1
			break
		}
	}

	if pos := binarySearch(nums, target, 0, splitPos); pos != -1 {
		return pos
	}

	return binarySearch(nums, target, splitPos+1, len(nums)-1)
}

func binarySearch(nums []int, target, left, right int) int {
	for left <= right {
		mid := (left + right) / 2
		switch {
		case nums[mid] == target:
			return mid

		case nums[mid] < target:
			left = mid + 1

		case nums[mid] > target:
			right = mid - 1
		}
	}

	return -1
}
